### ISSUE TITLE
Nested fragments not included in document (typescript-react-apollo with documentMode: documentNode)

### ISSUE THREAD

**Issue by adam1658** [dotansimha/graphql-code-generator#3564]
**Describe the bug**
When using typescript-react-apollo with `documentMode: documentNode`, nested fragments are not included in the generated documents.

**To Reproduce**
Try https://codesandbox.io/s/graphql-codegen-fragments-missing-98i12.
`yarn run codegen`
`yarn run test`

It produces a table that looks like this:
![image](https://user-images.githubusercontent.com/3508429/75499633-aa3f1100-5a1e-11ea-8308-1afcaf814b53.png)

This shows that it
- generated a document for the fragment `CFields`. It references no documents and defines no documents, as expected.
- generated a document for the fragment `BFields`. It references `CFields` and defines `CFields`, as expected.
- generated a document for the fragment `AFields`. It references `BFields` and `CFields` and defines `CFields`, **but it did not define `CFields`**
- generated a document for the query `Test`. It references `AFields` and `BFields` (but not `CFields`) and defines `AFields`, **but it did not define `BFields` or `CFields`**

1. My GraphQL schema:

```graphql
type Query {
  a: A
}

type A {
  bs: [B!]!
}

type B {
  cs: [C!]!
}

type C {
  greeting: String!
}
```

2. My GraphQL operations:

```graphql
query Test {
  a {
    ...AFields
  }
}

fragment AFields on A {
  bs {
    ...BFields
  }
}

fragment BFields on B {
  cs {
    ...CFields
  }
}

fragment CFields on C {
  greeting
}
```

3. My `codegen.yml` config file:

```yml
schema: schema.graphql
documents: document.graphql
generates:
  types.ts:
    plugins:
      - typescript
      - typescript-operations
      - typescript-react-apollo
    config:
      documentMode: documentNode
      withComponent: false
      withHOC: false
      withHooks: false
```

**Expected behavior**
I expected that the generated document for the `Test` query would include all the used fragments (`AFields`, `BFields` and `CFields`). Instead it only defined the `Test` query and the `AFields` fragment.

**Environment:**

- OS: linux (wsl2)
- `@graphql-codegen/typescript-react-apollo`: 1.12.2
- NodeJS: 12.16.1
- Also reproduced in linked codesandbox.

**Additional context**
Before I made the codesandbox I tried adding failing tests on a fork of graphql-code-generator, but I was a bit out of my depth. I can try again if that would be helpful and you can provide some guidance.

**Comment by pachuka** [comment]
I think this is a duplicate of #3066, but it is still an issue afaik.

**Comment by dotansimha** [comment]
You are right, thanks @adam1658 . Thank you for the super useful reproduction. 
It seems like when dealing with `documentNode` mode, we need to spread `SomeFragment.definitions` into the operation or the fragment that uses it. 
The goal is to make sure each fragment is defined once, and spread is being done to merge those. 

I'll take a look :)

**Comment by dotansimha** [comment]
With the upcoming `documentMode: documentNodeImportFragments` (available now with `1.12.3-alpha-d59d427f.102`) you'll be able to use it and have a similar outputs, but without duplications.

With `documentNodeImportFragments` you'll get in your test:

```
┌─────────┬────────────┬─────────────────────────────────────┬─────────────────────────────────────┐
│ (index) │ entrypoint │            usedFragments            │          definedFragments           │
├─────────┼────────────┼─────────────────────────────────────┼─────────────────────────────────────┤
│    0    │ 'CFields'  │                 []                  │                 []                  │
│    1    │ 'BFields'  │            [ 'CFields' ]            │            [ 'CFields' ]            │
│    2    │ 'AFields'  │      [ 'BFields', 'CFields' ]       │      [ 'BFields', 'CFields' ]       │
│    3    │   'Test'   │ [ 'AFields', 'BFields', 'CFields' ] │ [ 'AFields', 'BFields', 'CFields' ] │
└─────────┴────────────┴─────────────────────────────────────┴─────────────────────────────────────┘
```

In the case of regular `documentNode`, it should duplicate the fragments, and I fixed it in: https://github.com/dotansimha/graphql-code-generator/pull/3602 , the changes from that PR are available as alpha `1.12.3-alpha-81f0188d.104`, and using it with your test outputs the same (but in this case, it will get duplicated instead of JS spread for `definitions`):

```
┌─────────┬────────────┬─────────────────────────────────────┬─────────────────────────────────────┐
│ (index) │ entrypoint │            usedFragments            │          definedFragments           │
├─────────┼────────────┼─────────────────────────────────────┼─────────────────────────────────────┤
│    0    │ 'CFields'  │                 []                  │                 []                  │
│    1    │ 'BFields'  │            [ 'CFields' ]            │            [ 'CFields' ]            │
│    2    │ 'AFields'  │      [ 'BFields', 'CFields' ]       │      [ 'BFields', 'CFields' ]       │
│    3    │   'Test'   │ [ 'AFields', 'BFields', 'CFields' ] │ [ 'AFields', 'BFields', 'CFields' ] │
└─────────┴────────────┴─────────────────────────────────────┴─────────────────────────────────────┘
```
